stl/inc/thread

```diff
@@ -240,6 +254,93 @@ struct hash<thread::id> {
         return _Hash_representation(_Keyval._Id);
     }
 };
+
+#if _HAS_CXX20
+class jthread {
+public:
+    using id                 = thread::id;
+    using native_handle_type = thread::native_handle_type;
+
+    jthread() noexcept : _Impl{}, _Ssource{nostopstate} {}
+
+    template <class _Fn, class... _Args, enable_if_t<!is_same_v<_Remove_cvref_t<_Fn>, jthread>, int> = 0>
+    explicit jthread(_Fn&& _Fx, _Args&&... _Ax) {
+        if constexpr (is_invocable_v<decay_t<_Fn>, stop_token, decay_t<_Args>...>) {
+            _Impl._Start(_STD forward<_Fn>(_Fx), _Ssource.get_token(), _STD forward<_Args>(_Ax)...);
+        } else {
+            _Impl._Start(_STD forward<_Fn>(_Fx), _STD forward<_Args>(_Ax)...);
+        }
+    }
+
+    ~jthread() {
+        _Try_cancel_and_join();
+    }
+
+    jthread(const jthread&)            = delete;
+    jthread(jthread&& _Other) noexcept = default;
+    jthread& operator=(const jthread&) = delete;
+
+    jthread& operator=(jthread&& _Other) noexcept {
+        // note: the standard specifically disallows making self-move-assignment a no-op here
+        _Try_cancel_and_join();
+        _Impl    = _STD move(_Other._Impl);
+        _Ssource = _STD move(_Other._Ssource);
+        return *this;
+    }
+
+    void swap(jthread& _Other) noexcept {
+        _Impl.swap(_Other._Impl);
+        _Ssource.swap(_Other._Ssource);
+    }
+
+    [[nodiscard]] bool joinable() const noexcept {
+        return _Impl.joinable();
+    }
+
+    void join() {
+        _Impl.join();
+    }
+
+    void detach() {
+        _Impl.detach();
+    }

```

Comment:
I wonder whether this should be clearing out `_Ssource`?
The spec doesn't say anything about the state of the stop_source, so maybe this is fine.
Although it does say "When
the thread previously represented by `*this` ends execution, the implementation releases any owned
resources." - does this include resources used by the stop_source?

Reply:
My reading is that `detach` leaves `ssource` unchanged, so e.g. `request_stop` can still be used to request that the detached thread stop. This doesn't quite jibe with "When `detach()` returns, `*this` no longer represents the possibly continuing thread of execution.", which makes me think this may not be intentionally designed behavior.

Reply:
I don't think the `stop_source` is associated with the thread of execution at all; it's associated with this `std::jthread` object. In particular, the user can do:

```
stop_source src;
{
    jthread jthr {/* ... */}
    src = jthr.get_stop_source();
    jthr.detach();
}

src.request_stop();
```